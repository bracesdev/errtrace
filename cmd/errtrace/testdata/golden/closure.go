//go:build ignore

package foo

import (
	"errors"
	"fmt"
)

func ClosureReturnsError() error {
	return func() error {
		return errors.New("great sadness")
	}()
}

func ClosureDoesNotReturnError() error {
	x := func() int {
		return 42
	}()
	return nil
}

func DeferedClosureReturnsError() error {
	defer func() error {
		// The error is ignored,
		// but it should still be wrapped.
		return errors.New("great sadness")
	}()

	return nil
}

func DeferedClosureDoesNotReturnError() error {
	defer func() int {
		return 42
	}()

	return nil
}

func ClosureReturningErrorHasDifferentNumberOfReturns() (int, error) {
	x := func() error {
		return errors.New("great sadness")
	}

	return 42, x()
}

func ClosureNotReturningErrorHasDifferentNumberOfReturns() (int, error) {
	x := func() int {
		return 42
	}

	return 42, nil
}

func ClosureInsideAnotherClosure() error {
	return func() error {
		return func() error {
			return errors.New("great sadness")
		}()
	}()
}

func ClosureNotReturningErrorInsideAnotherClosure() (int, error) {
	var x int
	err := func() error {
		x := func() int {
			return 42
		}()
		return errors.New("great sadness")
	}()

	return x, err
}

func ClosureReturningAnErrorInsideADefer() error {
	defer func() {
		err := func() error {
			return errors.New("great sadness")
		}()

		fmt.Println(err)
	}()

	return nil
}

func ClosureNotReturningAnErrorInsideADefer() error {
	defer func() error {
		x := func() int {
			return 42
		}()

		return fmt.Errorf("great sadness: %d", x)
	}()

	return nil
}
