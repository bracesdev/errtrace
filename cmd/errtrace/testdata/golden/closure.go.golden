//go:build ignore

package foo

import (
	"errors"
	"fmt"; "braces.dev/errtrace"
)

func ClosureReturnsError() error {
	return errtrace.Wrap(func() error {
		return errtrace.Wrap(errors.New("great sadness"))
	}())
}

func ClosureDoesNotReturnError() error {
	x := func() int {
		return 42
	}()
	return nil
}

func DeferedClosureReturnsError() error {
	defer func() error {
		// The error is ignored,
		// but it should still be wrapped.
		return errtrace.Wrap(errors.New("great sadness"))
	}()

	return nil
}

func DeferedClosureDoesNotReturnError() error {
	defer func() int {
		return 42
	}()

	return nil
}

func ClosureReturningErrorHasDifferentNumberOfReturns() (int, error) {
	x := func() error {
		return errtrace.Wrap(errors.New("great sadness"))
	}

	return 42, errtrace.Wrap(x())
}

func ClosureNotReturningErrorHasDifferentNumberOfReturns() (int, error) {
	x := func() int {
		return 42
	}

	return 42, nil
}

func ClosureInsideAnotherClosure() error {
	return errtrace.Wrap(func() error {
		return errtrace.Wrap(func() error {
			return errtrace.Wrap(errors.New("great sadness"))
		}())
	}())
}

func ClosureNotReturningErrorInsideAnotherClosure() (int, error) {
	var x int
	err := func() error {
		x := func() int {
			return 42
		}()
		return errtrace.Wrap(errors.New("great sadness"))
	}()

	return x, errtrace.Wrap(err)
}

func ClosureReturningAnErrorInsideADefer() error {
	defer func() {
		err := func() error {
			return errtrace.Wrap(errors.New("great sadness"))
		}()

		fmt.Println(err)
	}()

	return nil
}

func ClosureNotReturningAnErrorInsideADefer() error {
	defer func() error {
		x := func() int {
			return 42
		}()

		return errtrace.Wrap(fmt.Errorf("great sadness: %d", x))
	}()

	return nil
}
